### PR TITLE
.Net: Unit tests for the MistralAI connector

### PR DESCRIPTION
**Pull request by markwallace-microsoft**
### Motivation and Context

Increase unit test coverage for the MistralAI connector

### Contribution Checklist

- [ ] The code builds clean without any errors or warnings
- [ ] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [ ] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/samples/Concepts/ChatCompletion/MistralAI_FunctionCalling.cs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        // Invoke chat prompt with auto invocation of functions enabled
+        const string ChatPrompt = @"
+            <message role=""user"">What is the weather like in Paris?</message>
+        ";
+        var executionSettings = new MistralAIPromptExecutionSettings { ToolCallBehavior = MistralAIToolCallBehavior.AutoInvokeKernelFunctions };
+        var chatSemanticFunction = kernel.CreateFunctionFromPrompt(
+            ChatPrompt, executionSettings);
+        var chatPromptResult = await kernel.InvokeAsync(chatSemanticFunction);
+
+        WriteLine(chatPromptResult);
```
Comment (RogerBarreto) [inline dotnet/samples/Concepts/ChatCompletion/MistralAI_FunctionCalling.cs:42]: Do for others too.
```suggestion
        Console.WriteLine(chatPromptResult);
```

File: dotnet/samples/Concepts/ChatCompletion/MistralAI_FunctionCalling.cs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+                modelId: TestConfiguration.MistralAI.ChatModelId!,
+                apiKey: TestConfiguration.MistralAI.ApiKey!,
+                httpClient: httpClient);
+        kernelBuilder.Plugins.AddFromType<WeatherPlugin>();
+        kernelBuilder.Plugins.AddFromType<WidgetFactory>();
+        Kernel kernel = kernelBuilder.Build();
+
+        // Invoke chat prompt with auto invocation of functions enabled
+        const string ChatPrompt = @"
+            <message role=""user"">Create a lime and scarlet colored widget for me.</message>
```
Comment (RogerBarreto) [inline dotnet/samples/Concepts/ChatCompletion/MistralAI_FunctionCalling.cs:160]: Use `"""` literals for better legibility

```suggestion
        const string ChatPrompt = """<message role="user">Create a lime and scarlet colored widget for me.</message>"""
```

File: dotnet/samples/Concepts/ChatCompletion/MistralAI_FunctionCalling.cs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+        WriteLine(chatPromptResult);
+    }
+
+    public sealed class WeatherPlugin
+    {
+        [KernelFunction]
+        [Description("Get the current weather in a given location.")]
+        public string GetWeather(
+            [Description("The city and department, e.g. Marseille, 13")] string location
+            ) => "12°C\nWind: 11 KMPH\nHumidity: 48%\nMostly cloudy";
```
Comment (RogerBarreto) [inline dotnet/samples/Concepts/ChatCompletion/MistralAI_FunctionCalling.cs:176]: nit.
```suggestion
        ) => "12°C\nWind: 11 KMPH\nHumidity: 48%\nMostly cloudy";
```

File: dotnet/samples/Concepts/ChatCompletion/OpenAI_FunctionCalling.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        WriteLine(result2[0].Content);
+    }
+
+    public sealed class WeatherPlugin
+    {
+        [KernelFunction]
+        [Description("Get the current weather in a given location.")]
+        public string GetWeather(
+            [Description("The city and department, e.g. Marseille, 13")] string location
+            ) => "12°C\nWind: 11 KMPH\nHumidity: 48%\nMostly cloudy";
```
Comment (RogerBarreto) [inline dotnet/samples/Concepts/ChatCompletion/OpenAI_FunctionCalling.cs:80]: nit.
```suggestion
         ) => "12°C\nWind: 11 KMPH\nHumidity: 48%\nMostly cloudy";
```

File: dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs [hunk 5]
```diff
@@ -238,117 +246,236 @@ await foreach (var chunk in response)
...
-            [Description("The city and department, e.g. Marseille, 13")] string location,
-            [Description("The temperature units one of celsius or fahrenheit")] TemperatureUnit units
-            ) => $"{{\"location\": \"{location}\", \"unit\": \"{units}\"}}";
+            [Description("The city and department, e.g. Marseille, 13")] string location
+            ) => "12°C\nWind: 11 KMPH\nHumidity: 48%\nMostly cloudy";
+    }
+
+    internal enum TemperatureUnit { Celsius, Fahrenheit }
+
+    public class WidgetFactory
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs:439]:
```suggestion
    private sealed class WidgetFactory
```

File: dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs [hunk 6]
```diff
@@ -238,117 +246,236 @@ await foreach (var chunk in response)
...
+        [Description("Creates a new widget of the specified type and colors")]
+        public string CreateWidget([Description("The colors of the widget to be created")] WidgetColor[] widgetColors)
+        {
+            var colors = string.Join('-', widgetColors.Select(c => c.GetDisplayName()).ToArray());
+            return $"Widget created with colors: {colors}";
+        }
+    }
+
+    [JsonConverter(typeof(JsonStringEnumConverter))]
+    public enum WidgetColor
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs:451]:
```suggestion
    private enum WidgetColor
```

File: dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs [hunk 7]
```diff
@@ -238,117 +246,236 @@ await foreach (var chunk in response)
...
+
+        [Description("Use when creating a green item.")]
+        Green,
+
+        [Description("Use when creating a blue item.")]
+        Blue
     }
 
-    public enum TemperatureUnit { Celsius, Fahrenheit }
+    internal sealed class FakeFunctionFilter(
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs:463]: I assume you are not using this subclass outside of this test, make all private.

```suggestion
    private sealed class FakeFunctionFilter(
```

File: dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs [hunk 8]
```diff
@@ -238,117 +246,236 @@ await foreach (var chunk in response)
...
+        Func<FunctionInvocationContext, Func<FunctionInvocationContext, Task>, Task>? onFunctionInvocation = null) : IFunctionInvocationFilter
+    {
+        private readonly Func<FunctionInvocationContext, Func<FunctionInvocationContext, Task>, Task>? _onFunctionInvocation = onFunctionInvocation;
+
+        public Task OnFunctionInvocationAsync(FunctionInvocationContext context, Func<FunctionInvocationContext, Task> next) =>
+            this._onFunctionInvocation?.Invoke(context, next) ?? Task.CompletedTask;
+    }
 
+    internal sealed class FakeAutoFunctionFilter(
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs:472]:
```suggestion
    private sealed class FakeAutoFunctionFilter(
```

File: dotnet/src/Connectors/Connectors.MistralAI.UnitTests/MistralTestBase.cs [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.SemanticKernel.Connectors.MistralAI.Client;
+using Microsoft.SemanticKernel.Http;
+using Xunit;
+
+namespace SemanticKernel.Connectors.MistralAI.UnitTests;
+public abstract class MistralTestBase : IDisposable
```
Comment (RogerBarreto) [inline dotnet/src/Connectors/Connectors.MistralAI.UnitTests/MistralTestBase.cs:14]: Have you considered adding this to TestUtilities ? 

We have in the `/InternalUtilities/samples/InternalUtilities/BaseTest.cs` may be interesting to bring another copy to Unit Tests.

```
  <ItemGroup>
    <Compile Include="$(RepoRoot)/dotnet/src/InternalUtilities/test/AssertExtensions.cs" Link="%(RecursiveDir)%(Filename)%(Extension)" />
    <Compile Include="$(RepoRoot)/dotnet/src/InternalUtilities/test/BaseTest.cs" Link="%(RecursiveDir)%(Filename)%(Extension)" />
  </ItemGroup>
```

**Review by dmytrostruk — APPROVED** [review]

**Review by RogerBarreto — APPROVED** [review]

File: dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs (hunk 5 above)
Comment (markwallace-microsoft) [reply dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs:439]: Causes potentially uninstatiated class erros

File: dotnet/src/Connectors/Connectors.MistralAI.UnitTests/MistralTestBase.cs (hunk 9 above)
Comment (markwallace-microsoft) [reply dotnet/src/Connectors/Connectors.MistralAI.UnitTests/MistralTestBase.cs:14]: I'll look into this in a sperate PR

File: dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs (hunk 6 above)
Comment (markwallace-microsoft) [reply dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs:450]: Will look at the visibility of the text classes and enums in a later PR

File: dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs (hunk 5 above)
Comment (RogerBarreto) [reply dotnet/src/Connectors/Connectors.MistralAI.UnitTests/Client/MistralClientTests.cs:438]: If the test works with them in private, pragma ignore them, 
The analyzer can't predict some class usages like serialization.

**Pull request merged**